### PR TITLE
docs: rewrite TESTING.md

### PR DESCRIPTION
**Pull request by DominicBurkart**
## Summary\n\n- Rewrites TESTING.md from scratch per #106: deletes the old 329-line guide and replaces it with a 245-word, mermaid-first doc.\n- Adds a flowchart showing the test topology (unit / integration / shell / eval -> CI).\n- Includes a quick-reference table linking every test category to its path and runner, plus actionable steps for adding new tests.\n\nCloses #106\n\nhttps://claude.ai/code/session_016LGnihh2HtPWjxCwUsU9Mz

### CODE REVIEW

**Review by DominicBurkart — COMMENTED** [review]
## Review of PR #121 -- docs: rewrite TESTING.md

The rewrite successfully condenses the old 329-line doc into a concise, mermaid-first reference. The flowchart and quick-reference table are useful. However, three of the five items in #106 are missing from this PR.

### Missing from #106

1. **Codecov gate** (#106 step 3): The issue requires documenting that codecov patch blocks PRs with <100% coverage. This is not mentioned anywhere in the new doc. Add a sentence or row to the table.

2. **Scope delineation one-liner** (#106 step 4): The issue asks for a one-line set of links clarifying: evals = agent performance (outcomes), tests = harness correctness (deterministic components), benches = resource optimization, QA = manual in dev env by an agent. This is absent. Add it near the top, before or after the mermaid diagram.

3. **ARCHITECTURE.md cross-link** (#106 step 5): The issue asks to link to the new TESTING.md from the mermaid venn in ARCHITECTURE.md (per #101). This PR only touches TESTING.md. Either add that change here or split it into a follow-up and note it in the PR body.

### Missing from plaidoire

The plaidoire in #106 contains actionable testing philosophy that the issue explicitly asks to encode as guidelines:
- Testing method priority order (realistic happy-path integration tests, doctests, proptests, unit tests, unhappy-path error legibility, edge-case doctests).
- Fail-fast principle: surface issues at compile time > CI > runtime.
- "Untested code is brittle. Harden it, don't bend it."
- Explicit prohibition of graceful degradation patterns (`|| true`, fallback params shadowing dead code, silent failures).

The current rewrite is purely structural (where tests live, how to run them). The issue asks for the guidelines to be based on this plaidoire. Consider adding a short "Testing principles" section (can be terse -- a bullet list is fine to stay under 300 words total).

### Accuracy

- The mermaid diagram and table are consistent with CLAUDE.md and the repo structure.
- Commands use `--all-features`, matching CLAUDE.md. Good.
- `cargo clippy` and `cargo fmt` lines match CLAUDE.md exactly.

### CI Status

All 17 check runs passed (unit x3 platforms, integration, integration-container, security, lint, builds, CI summary). No failures.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by DominicBurkart — COMMENTED** [review]
Addressed all four points from the review:

1. **Codecov gate** -- added a "CI gate" section explicitly stating patch coverage must be 100% and that a PR dropping below 100% is blocked.

2. **Scope delineation one-liner** -- added a **Scope:** line immediately after the heading (before the mermaid diagram): evals = agent performance (outcomes), tests = harness correctness (deterministic components), benches = resource optimization, QA = manual in dev env by an agent.

3. **ARCHITECTURE.md cross-link** -- added a "See [TESTING.md](TESTING.md) for the test topology..." line directly above the Container Topology mermaid diagram in ARCHITECTURE.md.

4. **Testing principles** -- added a "Principles" section drawn from the plaidoire: method priority order, fail-fast / compile-time > CI > runtime, and the explicit prohibition on graceful degradation (`|| true`, silent failures, fallback params shadowing bad/dead code).

TESTING.md stays within the 300-word target.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by DominicBurkart — COMMENTED** [review]
## Review of PR #121 — docs: rewrite TESTING.md

The rewrite is a strong improvement: concise, well-structured, and actionable for an agent. One factual error must be fixed before merge; two smaller issues are worth addressing.

---

### Bug (needs fix): CI gate coverage threshold is wrong

**"CI gate" section states:**
> Codecov patch coverage must be **100%** on every PR. A PR that drops patch coverage below 100% is blocked.

**`codecov.yml` at the head commit actually reads:**
```yaml
patch:
  default:
    # Relaxed from 100%: the #[ignore] e2e test and the
    # runtime-gated test_create_with_container_no_runtime exit early
    # in CI (no container runtime / Ollama available), so some new
    # container-dispatch branches are genuinely uncoverable there.
    target: 90%
```

The gate blocks at **90%**, not 100%. An agent reading this doc and writing new code that lands at 91% coverage will be confused when Codecov passes. Fix the stated threshold to 90% and keep the reason inline (container-dispatch branches are uncoverable in CI without a runtime).

---

### Issue (worth fixing): `cargo deny check` is listed locally but skipped in CI

The "Running tests" block includes:

```bash
# supply-chain
cargo deny check
```

`ci.yml` security job explicitly skips it:

```bash
echo "⚠️ Skipping cargo deny due to CVSS 4.0 incompatibility"
```

Listing a command that CI itself skips creates a misleading picture of what "passing" means. Either drop it from the local commands or add a note that it is currently disabled in CI and why.

---

### Minor: `image-builder` workspace member absent from "Where tests live" table

`Cargo.toml` lists three workspace members: `harness`, `model`, `image-builder`. The table covers `harness` and `model` but not `image-builder`. If `image-builder` intentionally has no tests, add a brief note so an agent doesn't treat the omission as an error of its own to investigate.

---

### Everything else is accurate

- Mermaid diagram nodes, commands, and file paths all match `ci.yml` and the actual repo layout.
- `--all-features` flag is consistent with CI unit and integration jobs.
- Clippy/fmt flags (`--workspace --all-targets --all-features -- -D warnings` and `--all -- --check`) match CI exactly.
- `model/tests/` exists and is correctly listed.
- `tests/security/`, `tests/integration/`, and `tests/lib/test-helpers.sh` all exist.
- Principles section accurately reflects the plaidoire from #106.
- ARCHITECTURE.md cross-link is present and correct.
- "Adding a new test" instructions are accurate for all three test categories.
- The 262 → 66 line reduction removes genuine bloat with no meaningful information loss.

---

_Generated by [Claude Code](https://claude.ai/code)_

---
_Generated by [Claude Code](https://claude.ai/code)_

File: TESTING.md [hunk 1]
```diff
@@ -1,262 +1,66 @@
...
-# Dependency checks
-./tests/security/test-dependencies.sh
-
-# Environment validation
-./tests/security/test-environment.sh
+## Principles
 
-# Security tool availability
-./tests/security/test-tools-availability.sh
+- Prefer in order: realistic happy-path integration tests, happy-path doctests on interfaces, proptests, unit tests (especially edge cases), unhappy-path error legibility, edge-case doctests.
```
Comment (DominicBurkart) [inline TESTING.md]:
```suggestion
- Sorted preferred strategies (note: you'll need an ensemble, use complementary approaches): realistic happy-path integration tests, happy-path doctests on interfaces, aneas/kani invariant validation, proptests, unit tests (edge cases), unhappy-path integration tests (error legibility), edge-case doctests. 
- Design interfaces which can use the highest standards for validation (example: a de/serialization path needs to be compatible with both integration tests and formal methods validation, an external API surface needs to be expressible with both doctests and proptests). 
```
